Add UrlSearchParams and unit tests

jason0x43 opened the pull request.

jason0x43 committed:
Add UrlSearchParams and unit tests

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+			list = {};
+		}
+
+		Object.defineProperty(this, 'list', { value: list });
```
Is this because list is private?  I'm not sure there's much point to it being non-enumerable.  @BryanForbes, want to weigh in?

jason0x43 replied:
It is.

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+	 * The internal list maps property keys to arrays of values. The value for any property that has been set will be an
+	 * array containing at least one item. Properties that have been deleted will have a value of 'undefined'.
+	 */
+	private list: { [key: string]: string[] };
```
I'd suggest underscore-prefixing this, and using `protected`.

kitsonk replied:
Or using core/WeakMap...

jason0x43 replied:
Hmmm...I thought we were trying to stay away from underscore prefixes as a privacy notation, but the style guide recommends it, so I suppose not.

jason0x43 replied:
WeakMap is great for privacy, but if we want to allow access in subclasses (which I would assume is why we'd switch from private to protected), it wouldn't work so well.

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+import { duplicate } from './lang';
+
+/**
+ * A object with string keys and string or string array values that describes a query string.
```
A -> An

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+type ParamList = { [key: string]: string | string[] };
+
+/**
+ * Parse a query string, returning a ParamList object.
```
Parse -> Parses (use third-person present for function summaries, as if it is describing what the function does)

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+ */
+function parseQueryString(input: string): ParamList {
+	const query = <{ [key: string]: string[] }> {};
+	input.split('&').forEach(entry => {
```
Consider `for...of` here instead.  (If we were using `forEach`, we'd avoid fat arrow here since there's no context preservation going on, as per the style guide.)

jason0x43 replied:
`for...of` seems like a good fit. I feel like our style guide is giving up one of the main benefits of arrow functions, though, in not recommending them for their compactness (which is kind of the whole point of the syntax).

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+}
+
+/**
+ * An class that represents a URL query search parameters.
```
An -> A (or consider dropping "A class that" entirely)

jason0x43 replied:
Apparently grammar had just totally abandoned me by the time I got to block comments.

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+ */
+export default class UrlSearchParams {
+	/**
+	 * Construct a new UrlSearchParams from a query string, an object of parameters and values, or anoter
```
Construct -> Constructs, anoter -> another

kfranqueiro commented on src/UrlSearchParams.ts:
```diff
@@ -0,0 +1,151 @@
+
+			const values = this.list[key];
+			const encodedKey = encodeURIComponent(key);
+			for (let i = 0; i < values.length; i++) {
```
Looks like you could use a `for...of` here.

jason0x43 replied:
I keep forgetting that exists...

kfranqueiro commented on tests/unit/UrlSearchParams.ts:
```diff
@@ -0,0 +1,130 @@
+				assert.deepEqual(params.getAll(''), [ 'foo', '' ]);
+			},
+
+			UrlSearchParams() {
```
Should we also have a test somewhere to verify that the two instances do not share values?

jason0x43 replied:
I added a new 'no data sharing' test for this.

jason0x43 committed:
UrlSearchParams updates based on feedback

- Clean up comments
- Rename `list` property to `_list`, make protected
- Add test to verify that copied instances don't share data

kfranqueiro commented:
Oops, I neglected to note this was merged in https://github.com/dojo/core/commit/6ec110dc25acc91952422d04d6a2e6c24b5c7c05.  Thanks, and sorry about that!
